### PR TITLE
chore: migrate to flat config

### PR DESCRIPTION
**Pull request by antfu**
Migrates to the new flat config: https://github.com/antfu/eslint-config

It should be able to cover more edge cases and improve the overall consistency.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -3,12 +3,44 @@
...
+  // Disable the default formatter, use eslint instead
+  "prettier.enable": false,
   "editor.formatOnSave": false,
-  "[json]": {
-    "editor.formatOnSave": false
+
+  // Auto fix
+  "editor.codeActionsOnSave": {
+    "source.fixAll.eslint": true,
+    "source.organizeImports": false
```
[the remark this answers is not in the document]
Comment (AriPerkkio) [reply .vscode/settings.json:17]: And I'm not sure why but I also need to add `"source.fixAll": "always"` here for getting autofix on save.

File: .vscode/settings.json [hunk 2]
```diff
@@ -3,12 +3,44 @@
     "birpc",
     "unmock"
   ],
+
+  // Enable the ESlint flat config support
+  "eslint.experimental.useFlatConfig": true,
+
+  // Disable the default formatter, use eslint instead
+  "prettier.enable": false,
   "editor.formatOnSave": false,
```
[the remark this answers is not in the document]
Comment (antfu) [reply .vscode/settings.json:12]: I am not so sure, shall we add that to every languages? Or maybe it's better for you to opt-in that for each project you want to do format on save?

File: .vscode/settings.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (antfu) [reply .vscode/settings.json:17]: Weird, I been using that for years, it works well and no errors. Did you have some extension checking it? Can you see where the error comes from?

File: test/core/test/__snapshots__/serialize.test.ts.snap [hunk 3]
```diff
@@ -19,7 +19,7 @@ exports[`error serialize > works 1`] = `
 {
   "array": [
     1,
-    ,
+    undefined,
```
Comment (sheremet-va) [inline test/core/test/__snapshots__/serialize.test.ts.snap:22]: Hm? What happened here?

**Review by sheremet-va — CHANGES_REQUESTED** [review]

**Comment by sheremet-va** [comment]
To be honest, this year there were _so_ many changes to eslint config, that I am a bit burned out. I would honestly just switch to prettier and forget about it. I don't have a strong preference here though.

**Comment by antfu** [comment]
If you believe it's good to switch to Prettier I am actually ok with it. Tho I doubt if the experience could be a lot better - we still need ESLint for the logics anyway. I am sorry about the frequent updates of my ESLint config, but I think most of them are not formatting related (so moving to prettier doesn't help). If some new rules don't make sense to Vitest feel free to disable them - or let me know and we could discuss for a better default.

**Comment by antfu** [comment]
This PR is created mostly for myself the checking the coverage of the new config, to me overall looks better and caught some inconsistency. Let's marked this draft until 1.0 of my ESLint config so it would be less trouble on maintenance.

**Comment by sheremet-va** [comment]
> If you believe it's good to switch to Prettier I am actually ok with it. Tho I doubt if the experience could be a lot better - we still need ESLint for the logics anyway. I am sorry about the frequent updates of my ESLint config, but I think most of them are not formatting related (so moving to prettier doesn't help). If some new rules don't make sense to Vitest feel free to disable them - or let me know and we could discuss for a better default.

Yes, moving to prettier doesn't really solve any of the problems. For me, the most important part is to have formatting consistent. _How_ the code is formatted doesn't really matter. Previously we had a lot of inconsistencies that were not caught by eslint. Eslint also wasn't applied to UI package for some reason 🤔 

If this update fixes it, I am all for it 👍🏻

**Comment by antfu** [comment]
Sure. My goal here is indeed to improve consistency. With the flat config is already getting much better, combined with the incentive of https://github.com/eslint-stylistic/eslint-stylistic it should get better over time.

File: .vscode/settings.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (AriPerkkio) [reply .vscode/settings.json:17]: VSCode update fixed this.

File: .vscode/settings.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (AriPerkkio) [reply .vscode/settings.json:12]: Sure, I can keep adding this myself as I've done so far. I just have to remember to revert this before opening PRs to upstream. 

I don't think there's a way for me to create global configuration that sets "use prettier everywhere except for <workspace-root>/vitest project".

**Pull request closed without merging**